examples_tests/10.AllocatorTest/main.cpp

```diff
@@ -0,0 +1,478 @@
+		for (uint32_t i = 0u; i < multiAllocCnt; i++)
+		{
+			outAddresses.clear();
+			sizes.clear();
+			alignments.clear();
+
+			// randomly decide how many allocs in a `multi_alloc` NOTE: must pick number less than `traits::max_multi_alloc`
+			constexpr uint32_t upperBound = core::address_allocator_traits<AlctrType>::maxMultiOps;
+			const uint32_t allocCntInMultiAlloc = rng.getRandomNumber(1u, upperBound);
+
+			for (size_t i = 0u; i < allocCntInMultiAlloc; ++i)
+			{
+				// randomly decide sizes (but always less than `address_allocator_traits::max_size`)
+				outAddresses.emplace_back(AlctrType::invalid_address);
+
+				if constexpr (std::is_same<AlctrType, core::PoolAddressAllocator<uint32_t>>::value)
+				{
+					sizes.emplace_back(randAllocParams.blockSz);
+					alignments.emplace_back(randAllocParams.blockSz);
+				}
+				else
+				{
+					sizes.emplace_back(rng.getRandomNumber(1u, Traits::max_size(alctr)));
+					alignments.emplace_back(rng.getRndMaxAlign());
+				}
+			}
+
+			Traits::multi_alloc_addr(alctr, allocCntInMultiAlloc, outAddresses.data(), sizes.data(), alignments.data());
+
+			// record all successful alloc addresses to the `core::vector`
+			for (uint32_t i = 0u; i < outAddresses.size(); i++)
+			{
+				if (outAddresses[i] != AlctrType::invalid_address)
+					results.push_back({ outAddresses[i], sizes[i], alignments[i] });
+			}
+
+			// run random dealloc function
+			// linear address allocator is always reseted here
+			randFreeAllocatedAddresses(alctr);
+		}
+
+		// randomly choose between reset and freeing all `core::vector` elements
+		if constexpr (!std::is_same<AlctrType, core::LinearAddressAllocator<uint32_t>>::value) //linear address allocator is always reseted in the `randFreeAllocatedAddresses()`
+		{
+			bool reset = static_cast<bool>(rng.getRandomNumber(0u, 1u));
+			if (reset)
+			{
+				alctr.reset();
+				results.clear();
+			}
+			else
+			{
+				// free everything with a series of multi_free
+				while (results.size() != 0u)
+					randFreeAllocatedAddresses(alctr);
+			}
+		}
+		 
+		if constexpr (!std::is_same<AlctrType, core::LinearAddressAllocator<uint32_t>>::value)
+			_IRR_ALIGNED_FREE(reservedSpace);
+
+	}
+
+	// random dealloc function
+	void randFreeAllocatedAddresses(AlctrType& alctr)
+	{
+		if(results.size() == 0u)
+			return;
+
+		outAddresses.clear();
+		sizes.clear();
+		alignments.clear();
+
+		// randomly decide how many calls to `multi_free`
+		const uint32_t multiFreeCnt = rng.getRandomNumber(1u, 10000u); //TODO
+
+		//TODO:
+		//shuffle results
+
+		for (uint32_t i = 0u; i < multiFreeCnt; i++)
+		{
+			// randomly how many addresses we should deallocate (but obvs less than all allocated) NOTE: must pick number less than `traits::max_multi_free`
+			const uint32_t addressesToFreeCnt = rng.getRandomNumber(0u, results.size()); //that should be restrained somehow, I think... so it it less likely to free all of allocated addresses
```

Comment:
upper bound should be `min(core::address_allocator_traits<AlctrType>::maxMultiOps,1)`

Reply:
dont restrain any more, we need to be able to hit mutli_free everything every once in a while, this could be an important edge case to test.